### PR TITLE
API: change the align default to None allowing silencing alignment warning in binary ops

### LINKED ISSUE #1832: API: default behaviour of alignment in binary ops
#1668 introduces `align` keyword to control whether two GeoSeries gets aligned in binary, row by row operations or not. However, the default behaviour may require a bit more discussion.

Moving from the original PR here (from https://github.com/geopandas/geopandas/pull/1668#discussion_r578692627).

> @jorisvandenbossche:  Idea: make the default None, which is treated as True but gives a warning when alignment is actually done. This gives a way for the user to indicate they are OK with alignment without getting warnings about it.
(small disadvantage is that the signature in the function is less explicit than with align=True)

> @martinfleis: I do understand your point, but considering how troublesome this behaviour is (I responded to countless questions on this), I would prefer having an explicit align=True in the signature. Having None/True/False is a nice trick for advanced users to silence warnings, but in this specific case, I would prefer understanding for basic users (I really don't want to answer the same question again :D).

> @jorisvandenbossche: Then we should maybe rather make it an error by default? To make it really clear for users they are doing something wrong (with a good explanation then)

> @martinfleis: That'd be a breaking change.

### PR DESCRIPTION
**Pull request by martinfleis**
Closes #1832 

One question here - do we want to test if the warning is emitted in every single binary method, since they all use the same `_delegate_binary_method` under the hood. I opted for _no_ but can potentially include those.

### CODE REVIEW

File: geopandas/base.py [hunk 1]
```diff
@@ -27,13 +27,20 @@ def is_geometry_type(data):
...
     if isinstance(other, GeoPandasBase):
         if align and not this.index.equals(other.index):
-            warn(
-                "The indices of the two GeoSeries are different.",
-                stacklevel=4,
-            )
+            if maybe_warn:
+                warn(
+                    "The indices of the two GeoSeries are different and will be "
+                    "aligned. Use the `align` keyword to control this behavior.",
```
Comment (jorisvandenbossche) [inline geopandas/base.py:41]: Do we want the expand the warning message as suggested in the issue?

**Comment by jorisvandenbossche** [comment]
I am certainly fine with not exhaustively testing this for every single method

File: geopandas/base.py (hunk 1 above)
Comment (martinfleis) [reply geopandas/base.py]: Forgot about that. Now in 84fff16

**Review by jorisvandenbossche — APPROVED** [review]
Looks good!

**Pull request merged**